File: Solutions/Cpp/same_tree.cpp
/*
    [EASY]
    100. Same Tree

    Concepts:
    - binary tree
    - breadth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 13.03 MB  [Beats 14.69%]
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        std::function<bool(TreeNode*, TreeNode*)> bfs;

        bfs = [](TreeNode* root1, TreeNode* root2){
            std::deque<TreeNode*> queue1 = {root1};
            std::deque<TreeNode*> queue2 = {root2};
            TreeNode* curr1;
            TreeNode* curr2;
            int n1, n2;

            while (!queue1.empty() && !queue2.empty()) {
                n1 = queue1.size();
                n2 = queue2.size();

                if (n1 != n2)
                    return false;
            
                for (int _ = 0; _ < n1; _++) {
                    curr1 = queue1.front();
                    curr2 = queue2.front();

                    queue1.pop_front();
                    queue2.pop_front();

                    if (
                        (curr1 != nullptr && curr2 == nullptr) ||
                        (curr2 != nullptr && curr1 == nullptr)
                    ) {
                        return false;
                    } else if (curr1 != nullptr && curr2 != nullptr) {
                        if (curr1->val != curr2->val)
                            return false;
                    }

                    if (curr1 != nullptr) {
                        queue1.push_back(curr1->left);
                        queue1.push_back(curr1->right);
                    }
                    if (curr2 != nullptr) {
                        queue2.push_back(curr2->left);
                        queue2.push_back(curr2->right);
                    }
                }
            }

            return true;
        };

        return bfs(p, q);
    }
};

File: Solutions/Rust/same_tree.rs
/*
    [EASY]
    100. Same Tree

    Concepts:
    - binary tree
    - breadth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 2.08 MB   [Beats 100%]
*/
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;

impl Solution {
    pub fn is_same_tree(
        p: Option<Rc<RefCell<TreeNode>>>,
        q: Option<Rc<RefCell<TreeNode>>>
    ) -> bool {
        fn bfs(
            root1: Option<Rc<RefCell<TreeNode>>>,
            root2: Option<Rc<RefCell<TreeNode>>>
        ) -> bool {
            let mut queue1: VecDeque<Option<Rc<RefCell<TreeNode>>>>
                = VecDeque::from([root1]);
            let mut queue2: VecDeque<Option<Rc<RefCell<TreeNode>>>>
                = VecDeque::from([root2]);
            let [mut n1, mut n2]: [usize; 2];
            let [mut curr1, mut curr2]: [Option<Rc<RefCell<TreeNode>>>; 2];

            while !queue1.is_empty() && !queue2.is_empty() {
                n1 = queue1.len();
                n2 = queue2.len();

                if n1 != n2 { return false; }

                for _ in 0..n1 {
                    if
                        let Some(curr1) = queue1.pop_front()
                        && let Some(curr2) = queue2.pop_front()
                    {
                        if
                            (curr1.is_some() && !curr2.is_some())
                            || (curr2.is_some() && !curr1.is_some())
                        {
                            return false;
                        }
                        if let Some(c1) = &curr1 && let Some(c2) = &curr2 {
                            if c1.borrow().val != c2.borrow().val { return false; }
                        }

                        for
                            (curr, queue)
                            in [(&curr1, &mut queue1), (&curr2, &mut queue2)]
                        {
                            if let Some(c) = curr {
                                queue.push_back(c.borrow().left.clone());
                                queue.push_back(c.borrow().right.clone());
                            }
                        }
                    }
                }
            }

            return true;
        }

        return bfs(p, q);
    }
}
